doc/reference/trd-storage-permissions.md

```diff
@@ -0,0 +1,183 @@
+This document is in full compliance with [TRD1][TRD1].
+
+
+1 Introduction
+-------------------------------
+
+Tock applications need to be able to store persistent state. Additionally,
+applications need to be able to keep data private from other applications. The
+kernel should also be able to allow specific applications to read and modify
+state from other applications.
+
+This requires a method for assigning applications persistent identifiers,
+a mechanism for granting storage permissions to specific applications,
+and kernel abstractions for implementing storage capsules that respect
+the storage permissions.
+
+
+2 Scope
+-------------------------------
+
+This document only describes the permission architecture in Tock for supporting
+application persistent storage. This document does not include specific types of
+persistent storage (e.g., flash, FRAM, etc.) or storage access abstractions
+(e.g., block-access, byte-access, etc.).
+
+
+3 Stored State Identifiers
+-------------------------------
+
+All shared persistent storage implementations must store a 32 bit identifier
+with each stored object to mark the application that created the stored object.
+
+When applications write data, their ShortId must be used as the identifier. When
+the kernel writes data, the identifier must be 0.
+
+
+4 Permissions
+-------------------------------
+
+All persistent application data is labeled based on the application which wrote
+the data.
+Applications can read and modify data with suitable
+permissions.
+
+There are three types of permissions:
+
+1. **Write**: The application can write data.
+1. **Read**: The application can read data.
+1. **Modify**: The application can modify existing data.
+
+Each permission type is independent. For example, an application can be given
+read permission for specific data but not be able to write new data itself.
+
+Write is a boolean permission. An application either has permission to write or
+it does not.
+
+Read and Modify permissions are tuples of `(the permission type, stored state
+identifier)`. These permissions only exist as associated with a particular
+stored state identifier. That is, a Read permission gives an application
+permission to read only stored state marked with the associated stored state
+identifier, and a Modify permission gives an application permission to modify
+only stored state marked with the associated stored state identifier.
+
+
+5 Requirements
+-------------------------------
+
+The Tock storage model imposes the following requirements:
+
+1. Applications are given separate write, read, and modify permissions.
+2. The label stored with the persistent data when the data are written is the
+   application's short AppID.
+3. Applications without a `ShortId::Fixed` cannot access (i.e.,
+   read/write/modify) any persistent storage.
+4. How permissions are mapped to applications must be customizable for different
+   Tock kernels.
```

Comment:
What does this mean?

I would think we would, at least to start, want a more limited/specific set of ways to give apps storage permissions?

e.g., my instinct was that putting storage permissions in TBF headers would be sufficient?

This feels a bit like under-specifying a hard problem here—maybe it's clearer below?

Reply:
This means that the design for storage permissions cannot be such that there is only one way to specify per-app permissions.

Using TBF header's doesn't work well because that allows applications to set their own permissions. Signing can help, but a kernel may want to support signed apps from untrusted developers.

I listed a couple ideas in the TRD. What is the benefit of trying to be prescriptive?

Reply:
Maybe just delete this bullet here?

Reply:
> Using TBF header's doesn't work well because that allows applications to set their own permissions. Signing can help, but a kernel may want to support signed apps from untrusted developers.

The solution to this was have the board `main()` verify that applications have permitted  headers on a case by case basis.

Reply:
> Maybe just delete this bullet here?

I think this is an important requirement. It basically motivates why we have a `StoragePermissions` trait rather than a struct. It makes explicit that the design we want is flexible (eg like the scheduler) rather than fixed (more like the grants implementation).

Reply:
This is specifying _how_ to specify the access control policy in Tock. There can (probably should) also be an implementation of this trait which is a recommended default.

Reply:
> The solution to this was have the board `main()` verify that applications have permitted headers on a case by case basis.

I don't understand how the kernel can do that. Lets say a Tock system that supports the dynamic application loading has the following two binaries installed:

| Name | App ID | Write ID | Read IDs | Modify IDs |
| ---- | ------ | -------- | -------- | ---------- |
| Alpha | AAAA | 1 | 2, 3 | 3, 4 |
| Bravo | BBBB | 2 | 1, 2, 3, 4 | |

Meanwhile, its storage layer contains the following files:

| Name | data ID |
| ---- | -------- |
| access_log.txt | 1 |
| auth_config.json | 2 |

It is then sent a command to install a third application. It downloads the TBF, and deploys the new image, which has the following permissions in its TBF header:

| Name | App ID | Write ID | Read IDs | Modify IDs |
| ---- | ------ | -------- | -------- | ---------- |
| Charlie | CCCC | 3 | 1, 2, 3, 4 | 1, 2, 3, 4 |

How can the kernel determine whether these permissions are permitted? Obviously this isn't enough data -- so the kernel has to look at *something else*. I'm struggling to see what the kernel has access to that it could use to decide whether Charlie's premissions are acceptable or unacceptable.

Is there something I'm missing?

Reply:
The kernel can use the signature chain. So for example only apps signed by company A can use ID 1 or 2. Apps signed by company B can use 3 and 4. The idea being that you can limit sensitive access to certain private keys. The same goes for syscall filtering

Reply:
> The kernel can use the signature chain. So for example only apps signed by company A can use ID 1 or 2. Apps signed by company B can use 3 and 4. The idea being that you can limit sensitive access to certain private keys. The same goes for syscall filtering

What signature chain?

If you're referring to the signature chain for cryptographically-verified app IDs, then that would require the kernel to store a table mapping app IDs to the storage IDs those app IDs have permission to use. If that's the case, then why have these TBF headers at all?

Reply:
With AppID, if the kernel maintains a table like:

| AppId | Acceptable WriteIds |
|-------|-------|
| AAA1 | 1, 2 |
| AAA2 | 1, 2 |
| BBB1 | 3, 4 |
| BBB2 | 3, 4 |

Then the kernel can enforce the mapping, but not have to decide which is the correct write ID to use. The app compiler can chose which write ID to use (via the headers), but the kernel maintainer can ensure that company A apps can only use certain write IDs and company B apps can only use certain write IDs.

(Note, I'm not saying we should have write IDs in TBF headers. This same functionality can be had by setting and verifying read and modify permissions. And using AppIds directly means not having to duplicate this functionality, ie once in assigning AppIds and then again in checking storage permissions.)

Reply:
With that table design, if a new app (CCC) is deployed that wants to read from storage ID 2, how does the kernel know that is acceptable? Who "owns" storage ID 2? AAA1? AAA2?

I think the table would have to look more like:

| Storage ID | Owning app ID | App IDs that can write | App IDs that can read | App IDs that can modify |
| - | - | - | - | - |
| 1 | AAA | AAA, BBB | AAA, BBB | AAA, BBB |
| 2 | AAA | AAA | AAA | AAA |
| 3 | BBB | BBB | AAA, BBB | BBB |

An additional problem is there's not a good way to generate these storage IDs. If two different app developers pick the same storage ID for one of their apps, then those apps become incompatible. This is a general problem with the use of developer-assigned 32-bit storage IDs -- it's not specific to this table design, the table design just makes the issue apparent.

Reply:
> With that table design, if a new app (CCC) is deployed that wants to read from storage ID 2, how does the kernel know that is acceptable? Who "owns" storage ID 2? AAA1? AAA2?

I think company A "owns" storage IDs 1 and 2. So an app CCC from company C would not be able to read storage ID 2. Assuming the header tried to add read id 2 for CCC, the kernel storage permissions policy would deny that request.

Now, if the goal _is_ to allow CCC to read A's "owned" storage ID 2, then I'm not sure how to do that in this model. If we go back a step or too, the kernel could maintain a different table:

| AppID | Trust Storage Permissions Header | Acceptable WriteIDs |
|----|------|----|
| AAA* | Yes | 1,2 |
| BBB* | Yes | 3,4 |
| CCC* | Yes | 5,6 |
| DDD* | No | 7,8 |
| * | No | |

WriteIDs are allocated per company in advance. Read/modify permissions are trusted if signed by a trusted company.

-------

I'm lost as to what we are actually discussing here. This TRD does not prevent (nor mandate) a kernel from supporting a Storage Permission TBF header. The only restriction is the StorageID comes from AppID and not from any other source. If a kernel wants to maintain any of these hypothetical tables it can. The only difference between this hypothetical discussion and what is implementable per this TRD is that the kernel storage permissions policy may have to allocate additional read and modify permissions.

Reply:
I'm responding to this statement from Alistair:

> > Using TBF header's doesn't work well because that allows applications to set their own permissions. Signing can help, but a kernel may want to support signed apps from untrusted developers.
> 
> The solution to this was have the board `main()` verify that applications have permitted headers on a case by case basis.

I simply don't see how that is possible in a way that doesn't do one of the following:

1. Make storage IDs effectively aliases for app IDs (in which case storage IDs seem redundant)
2. Substantially duplicate the app ID infrastructure (e.g. have two parallel cryptographic signing systems).

Reply:
> What signature chain?

You can sign the TBF object (header and binary) with a certificate. That certificate can be signed by other certificates creating a chain, which can be stored in the TBF footer.

The kernel can then include details on trusted rootCAs (a serial number to compare against would be enough) and compare against that to verify if an application is signed by a trusted party.

> If you're referring to the signature chain for cryptographically-verified app IDs, then that would require the kernel to store a table mapping app IDs to the storage IDs those app IDs have permission to use. If that's the case, then why have these TBF headers at all?

I don't think the kernel would use appIDs for this. The main part is checking the rootCA that is used to sign the binary and TBF header.

The TBF headers can still reduce scope. So if company Foo creates the kernel, only apps signed by company Foo rootCA can access read/write IDs 1, 2 and 3.

Company Foo can then create multiple apps, each one maybe only using a single read/write ID for isolation. Third party apps can then using read/write ID 4.

Reply:
> I don't think the kernel would use appIDs for this. The main part is checking the rootCA that is used to sign the binary and TBF header.

If a kernel isn't using AppId (the system described in the appid trd), but is verifying signatures, at some point it is quite different from upstream tock. This storage TRD, at the end of the day, is just some text living on a server somewhere. It might be reasonable at that point to just ignore this TRD. This TRD is based on AppId being the method for cryptographically verifying applications.

Reply:
I said I don't think the kernel will use app IDs to map storage IDs to permissions. App IDs can be used elsewhere